### DISCUSSION TITLE
Bug: Request has no attribute 'htmx'

### DISCUSSION THREAD

**Discussion by yujinyuz** [litestar-org/litestar#3724] [Q&A] [answered]
### Description

I'm trying out litestar with htmx and I encountered this problem while following the documentation

`AttributeError: 'Request' object has no attribute 'htmx'

### MCVE

```python
from pathlib import Path
from litestar import get, Litestar
from litestar.response import Template
from litestar.template import TemplateConfig

from litestar.contrib.jinja import JinjaTemplateEngine
from litestar.contrib.htmx.request import HTMXRequest
from litestar.contrib.htmx.response import HTMXTemplate


@get("/")
async def index() -> Template:
    return Template(template_name="pages/index.html")


@get(path="/form")
async def get_form(request: HTMXRequest) -> Template:
    htmx = request.htmx  #  if true will return HTMXDetails class object
    if htmx:
        print(htmx.current_url)

    context = {}
    return HTMXTemplate(
        template_name="partials/sample.html",
        context=context,
        push_url="/form",
    )


template_config = TemplateConfig(
    directory=Path(__file__).parent / "templates",
    engine=JinjaTemplateEngine,
)


app = Litestar(
    route_handlers=[index, get_form],
    template_config=template_config,
)
```

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Simply HTMX</title>
  <link href="" rel="stylesheet">

  <script src="https://unpkg.com/htmx.org"></script>
</head>

<body>
  <button id="test-button" class="button is-primary" hx-get="/form" hx-swap="outerHTML">
    I'm new here...
  </button>
</body>

</html>
```

### Steps to reproduce

```bash
1. Run the script or the sample repository here https://github.com/yujinyuz/simplystar
2. Click the button
3. You should get the error
```

### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)"
```

### Litestar Version

2.11.0

### Platform

- [ ] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3723">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3723/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/3723/pledge.svg">
</picture>
</a>

**Comment by Alc-Alc** [comment, marked as answer]
You will have to specify `request_class=HTMXRequest` at the app or route handler layer. This is not a bug.

**Reply by provinzkraut** [reply]
@Alc-Alc what do you think about extracting the `request_class` from the `request` annotation?

**Reply by Alc-Alc** [reply]
> @Alc-Alc what do you think about extracting the `request_class` from the `request` annotation?

I would like to know how this would work with the layered nature of this feature. Are you suggesting the implicit conversion (from the function arg) will be in addition to the layered config we have?

**Reply by provinzkraut** [reply]
I think it should behave just like setting `request_class` on the route handler. That is, it should override a `request_class` on the layers defined above. I don't think we want to expand this to being able to extract this information from e.g. dependencies as well.
